Rubberduck.VBEEditor/ComManagement/TypeLibsAPI.cs

```diff
@@ -0,0 +1,114 @@
+                    .StdModExecute(procName, Reflection.BindingFlags.InvokeMethod, args);
+            }
+        }
+
+        // returns the raw conditional arguments string, e.g. "foo = 1 : bar = 2"
+        public static string GetProjectConditionalCompilationArgsRaw(IVBE ide, string projectName)
+        {
+            using (var typeLibs = new TypeLibsAccessor_VBE(ide))
+            {
+                return typeLibs.FindTypeLib(projectName).ConditionalCompilationArguments;
+            }
+        }
+
+        // return the parsed conditional arguments string as a Dictionary<string, string>
+        public static Dictionary<string, string> GetProjectConditionalCompilationArgs(IVBE ide, string projectName)
+        {
+            var args = GetProjectConditionalCompilationArgsRaw(ide, projectName);
+
+            if (args.Length > 0)
+            { 
+                string[] argsArray = args.Split(new[] { ':' });
+                return argsArray.Select(item => item.Split('=')).ToDictionary(s => s[0], s => s[1]);
+            }
+            else
+            {
+                return new Dictionary<string, string>();
+            }
+        }
+
+        // sets the raw conditional arguments string, e.g. "foo = 1 : bar = 2"
+        public static void SetProjectConditionalCompilationArgsRaw(IVBE ide, string projectName, string newConditionalArgs)
+        {
+            using (var typeLibs = new TypeLibsAccessor_VBE(ide))
+            {
+                typeLibs.FindTypeLib(projectName).ConditionalCompilationArguments = newConditionalArgs;
+            }
+        }
+
+        // sets the conditional arguments string via a Dictionary<string, string>
+        public static void SetProjectConditionalCompilationArgs(IVBE ide, string projectName, Dictionary<string, string> newConditionalArgs)
+        {
+            var rawArgsString = string.Join(" : ", newConditionalArgs.Select(x => x.Key + " = " + x.Value));
+            SetProjectConditionalCompilationArgsRaw(ide, projectName, rawArgsString);
+        }
+
+        public static bool IsExcelWorkbook(IVBE ide, string projectName, string className)
+        {
+            return DoesClassImplementInterface(ide, projectName, className, "Excel", "_Workbook");
+        }
+
+        public static bool IsExcelWorksheet(IVBE ide, string projectName, string className)
+        {
+            return DoesClassImplementInterface(ide, projectName, className, "Excel", "_Worksheet");
+        }
+
+        public static bool DoesClassImplementInterface(IVBE ide, string projectName, string className, string typeLibName, string interfaceName)
+        {
+            using (var typeLibs = new TypeLibsAccessor_VBE(ide))
+            {
+                return typeLibs.FindTypeLib(projectName).FindTypeInfo(className).DoesImplement(typeLibName, interfaceName);
+            }
+        }
+
+        public static bool DoesClassImplementInterface(IVBE ide, string projectName, string className, Guid interfaceIID)
```

Comment:
Great, much better now that we can fully qualify the name. A question, though. Within VBA code, we get to say `Excel.WorkBook` but in your sample, we have to check `Excel` and `_Workbook`. Do we need a helper method to translate between the ProgID and the actual interfaces being used?

Reply:
Probably need to avoid doing a prog-id lookup for each call to us here (as a registry lookup is costly).  But sure, a helper somewhere to get the root interface from a class prog-id would be helpful to expose.

Reply:
Be careful though, as the implementations of document classes can be messy.  For example, if you run a project in Access 2000 you it might inherit from Access._Form, but then in later versions its Access._Form2 or _Form3.  And as you can have multiple versions of the host installed, who says the registry entry will be right?   You may unfortunately need to hardcode the interface names sometimes.

Reply:
^ IMO in Access they made a mistake there, as _Form3 should have inherited from _Form2 etc.  But we'll have to make do.

Reply:
GTK. We might need to try to avoid on expecting a certain interface but rather asking if it can accept one of those possible interfaces, then or a similar approach for scenarios where we don't know the interface at the hand.

Reply:
Yes that'll be most useful for testing casting of objects at compile-time or something similar.